oxidation/libparsec/crates/core_fs/src/storage/local_database.rs

```diff
@@ -1,48 +1,150 @@
+    pub fn new<P: Into<String>>(path: P) -> FSResult<Self> {
         let path = path.into();
         if let Some(prefix) = PathBuf::from(&path).parent() {
             std::fs::create_dir_all(prefix).map_err(|_| FSError::CreateDir)?;
         }
-        let manager = ConnectionManager::<SqliteConnection>::new(path);
-        Pool::builder()
-            .build(manager)
-            .map_err(|_| FSError::Pool)
-            .map(Self)
-    }
-
-    pub fn conn(&self) -> FSResult<SqliteConn> {
-        let mut conn = self
-            .0
-            .get()
-            .map_err(|e| FSError::Connection(e.to_string()))?;
+
+        let mut connection =
+            SqliteConnection::establish(&path).map_err(|e| FSError::Connection(e.to_string()))?;
+
         // The combination of WAL journal mode and NORMAL synchronous mode
         // is a great combination: it allows for fast commits (~10 us compare
         // to 15 ms the default mode) but still protects the database against
         // corruption in the case of OS crash or power failure.
-        conn.batch_execute(
-            "
+        connection
+            .batch_execute(
+                "
             PRAGMA journal_mode = WAL; -- better write-concurrency
             PRAGMA synchronous = NORMAL; -- fsync only in critical moments
         ",
-        )
-        .map_err(|_| FSError::Configuration)?;
+            )
+            .map_err(|_| FSError::Configuration)?;
+
+        Ok(connection.into())
+    }
+
+    /// Get exclusive access to the Sqlite connection among cloned [SqliteConn]s.
+    pub fn lock_connection(&self) -> LockedSqliteConnection {
+        self.connection.lock().expect("Mutex is poisoned").into()
+    }
+
+    /// Close the connection to the local database.
+    pub fn close_connection(&self) {
+        *self.lock_connection().conn = None;
+    }
+
+    #[inline]
+    pub fn exec<T, F>(&self, f: F) -> Result<T, ExecError>
+    where
+        F: FnOnce(&mut SqliteConnection) -> Result<T, diesel::result::Error>,
+    {
+        self.lock_connection().exec(f)
+    }
+
+    #[inline]
+    pub fn exec_with_handler<T, F, G>(&self, f: F, error_handler: G) -> FSResult<T>
+    where
+        F: FnOnce(&mut SqliteConnection) -> Result<T, diesel::result::Error>,
+        G: FnOnce(diesel::result::Error) -> FSError,
+    {
+        self.lock_connection().exec_with_handler(f, error_handler)
+    }
+}
+
+impl From<SqliteConnection> for SqliteConn {
+    fn from(conn: SqliteConnection) -> Self {
+        Self {
+            connection: Arc::new(Mutex::new(Some(conn))),
+        }
+    }
+}
+
+pub struct LockedSqliteConnection<'a> {
+    conn: MutexGuard<'a, Option<SqliteConnection>>,
+}
+
+impl<'a> From<MutexGuard<'a, Option<SqliteConnection>>> for LockedSqliteConnection<'a> {
+    fn from(guard: MutexGuard<'a, Option<SqliteConnection>>) -> Self {
+        Self { conn: guard }
+    }
+}
+
+impl<'a> LockedSqliteConnection<'a> {
+    #[inline]
```

Comment:
Why do you want to force `inlining` ?

Reply:
I don't force the `inlining` of the method see <https://doc.rust-lang.org/reference/attributes/codegen.html#the-inline-attribute>

Why I use `inline` 
> Before this PR the logic to access the wrapped connection was present on every object that used `SqliteConn`

Reply:
PS: It may be better to let the compiler judge if this method is best to be inlined or not, I will remove this attribute

Reply:
The request is >> inlining optimization...